mlir/lib/Dialect/Rock/Transforms/FindFirstGemmIndex.cpp

```
@@ -59,45 +61,97 @@ static bool canTraceToFirstGemmArg(Value input, Value firstGemmArg) {
+      gemmGemmOp.getPreSecondGemmRegion().getArgument(firstGemmIndices[0]);
+
+  BufferDependencyAnalysis analysis(func.getOperation());
+  /*
+  genOpsList is list of linalg.generic ops in the preSecondGemmRegion in the
+  order they appear after the first gemm
+  */
+  for (auto genOp : genOpsList) {
+    SmallVector<int64_t> newFirstGemmIndices;
+    for (auto [index, input] : llvm::enumerate(genOp.getInputs())) {
+      if (canTraceToFirstGemmArg(input, firstGemmOutArg)) {
+        // If the input can be traced back to the first gemm argument, we found
+        // the new index.
+        newFirstGemmIndices.push_back(index);
+      } else {
+        // trace input to any of the previously visited linalg.generic ops
+        SmallVector<rock::TransformOp> transformOps;
+        Value untransformed;
+        std::tie(untransformed, std::ignore) =
+            rock::untransform(input, transformOps);
+        if (memref::AllocOp allocOp =
+                untransformed.getDefiningOp<memref::AllocOp>()) {
+          std::optional<llvm::SmallVector<OpOperand *>> writers =
+              analysis.getWriters(allocOp);
+          if (writers == std::nullopt) {
+            // If the allocOp has no writers, we cannot trace it back to a
+            // previous linalg.generic op.
+            continue;
+          }
+          // else
+          // If the allocOp has writers, check if any of them is a
+          // linalg.generic op that we have already visited.
+          for (OpOperand *writer : writers.value()) {
+            if (auto writerOp =
+                    dyn_cast<linalg::GenericOp>(writer->getOwner())) {
+              // If the writer is a linalg.generic op, check if it is in the
+              // genOpsList and if it has been processed already.
+              if (genOpToIndexMap.contains(writerOp))
+                newFirstGemmIndices.push_back(index);
+            }
+          }
+        }
+      }
+    }
+    if (newFirstGemmIndices.empty()) {
```

I think we might find some linalgs that don't trace back to the first GEMM output (although they are probably input fusions that are placed here). I think handling those is easy, just have a sentinel value (-1 for example), to indicate no index traces back to the first GEMM.

---

> although they are probably input fusions that are placed here

Yes, If none of the input traces back to gemm then it should fail. Linalg op would be placed incorrectly inside preSoftmaxBody. It should have been part of input fusion. It is better to fail as it would be invalid IR for further processing.